feat: upgrade to Olive

Agrendalath opened the pull request:
This upgrades the Simple Theme to Olive.

TODO:
- [ ] Bump version.
- [ ] Add comments to the PR explaining these changes.

## Configuration
```scss
$main-color: violet;  
$link-color: green;  
$header-bg: yellow;  
$footer-bg: aqua;  
  
$customize-header-border: true;  
$header-border-top-width: 10px;  
$header-border-top-color: red;  
$header-border-bottom-width: 8px;  
$header-border-bottom-color: blue;  
  
$customize-footer-border: true;  
$footer-border-top-width: 7px;  
$footer-border-top-color: blue;  
$footer-border-bottom-width: 11px;  
$footer-border-bottom-color: red;  
  
$btn-primary-bg: red;  
$btn-primary-color: gold;  
$btn-primary-border-color: black;  
$btn-primary-hover-bg: forestgreen;  
$btn-primary-hover-color: honeydew;  
$btn-primary-hover-border-color: dodgerblue;  
  
$btn-secondary-bg: forestgreen;  
$btn-secondary-color: honeydew;  
$btn-secondary-border-color: dodgerblue;  
$btn-secondary-hover-bg: red;  
$btn-secondary-hover-color: gold;  
$btn-secondary-hover-border-color: black;  
  
$main-nav-color: firebrick;  
$main-nav-link-color: lawngreen;  
$main-nav-item-border-bottom-color: mediumpurple;  
$main-nav-item-hover-border-bottom-color: yellowgreen;  
$user-dropdown-color: plum;  
  
$customize-sign-in-btn: true;  
$btn-sign-in-bg: cyan;  
$btn-sign-in-color: magenta;  
$btn-sign-in-border-color: green;  
$btn-sign-in-hover-bg: salmon;  
$btn-sign-in-hover-color: royalblue;  
$btn-sign-in-hover-border-color: red;  
  
$customize-register-btn: true;  
$btn-register-bg: salmon;  
$btn-register-color: royalblue;  
$btn-register-border-color: red;  
$btn-register-hover-bg: cyan;  
$btn-register-hover-color: magenta;  
$btn-register-hover-border-color: green;  
  
$customize-logistration-action-btn: true;  
$btn-logistration-bg: saddlebrown;  
$btn-logistration-color: purple;  
$btn-logistration-border-color: turquoise;  
$btn-logistration-hover-bg: thistle;  
$btn-logistration-hover-color: tomato;  
$btn-logistration-hover-border-color: wheat;  
  
$login-register-header-color: red;  
$accent-color: purple;  
  
$home-page-hero-title-color: darkorange;  
$home-page-hero-subtitle-color: blue;  
  
$wrapper-preview-menu-color: rebeccapurple;  
  
$course-nav-menu-color: green;  
$course-nav-menu-border-bottom-color: red;  
  
$footer-color: orangered;  
$footer-link-color: fuchsia;
```

## Instructions
1. Set up the theme in the legacy frontend as described [here](https://docs.tutor.overhang.io/tutorials/theming.html).
2. Clone this repository into the MFE repository (otherwise it won't be accessible from tutor).
3. Run `npm i && npm i @edx/brand@file:./edx-simple-theme/` in the MFE repository.
4. Mount the MFE with `tutor dev start --mount=. learning`. (change `learning` to `account` and `profile` to use the theme with other MFEs).

### Omitted parts
1. "Staff" and "Studio" buttons on the staff navbar. Their class is `btn-inverse-outline-primary"`, but inverting the primary button colors didn't look right here. If we want to override them, we should think about introducing new variables.
2. "Example handout". It is rendered in an iframe.

### Screenshots

#### Course Outline (learning MFE):
![Pasted image 20230131180958](https://user-images.githubusercontent.com/3189670/215857460-345238cf-b4c3-4242-9e2e-3337ac04f736.png)

#### Courseware:
![Pasted image 20230131182118](https://user-images.githubusercontent.com/3189670/215857459-3201bd4e-fe30-4340-ad2b-16330834ffe7.png)

#### Courseware (Legacy: Preview mode)
![Pasted image 20230131182210](https://user-images.githubusercontent.com/3189670/215857451-eafa4d7a-006f-43ae-ba38-4693d08858c1.png)

#### Course About page (Legacy)
![Pasted image 20230131182423](https://user-images.githubusercontent.com/3189670/215857447-13457407-f9b5-487d-a1f9-4bc4a8725eb4.png)

#### Learner Dashboard (Legacy)
![Pasted image 20230131182509](https://user-images.githubusercontent.com/3189670/215857445-2dbde3ee-dbc6-453f-b531-0f44f5ae3306.png)

#### Instructor Dashboard (Legacy)
![Pasted image 20230131182605](https://user-images.githubusercontent.com/3189670/215857441-27402c07-1752-47cf-86be-71250b4cf913.png)

#### Profile MFE
![Pasted image 20230131184642](https://user-images.githubusercontent.com/3189670/215857438-f8f29cb4-3477-43b5-b2bb-70af3b9c90be.png)

#### Account MFE
![Pasted image 20230131191945](https://user-images.githubusercontent.com/3189670/215857435-389d6866-ae44-4cec-869f-f930d97c4ec9.png)

Private-ref: [BB-7050](https://tasks.opencraft.com/browse/BB-7050)

Agrendalath commented on common/_overrides.scss:
```diff
@@ -19,11 +19,12 @@
-a:not(.btn):focus,
-a:visited:not(.btn):hover,
-a:visited:not(.btn):focus {
+// HACK: The first rule splits classes into separate :not selectors to increase the specificity.
+a[href]:not(.btn):not(.help-link),
+a[href]:not(.btn):hover,
+a[href]:not(.btn):focus,
+a[href]:visited:not(.btn):hover,
+a[href]:visited:not(.btn):focus {
```
E.g., the Account MFE is a selector with higher specificity at the bottom of the page. While we could add `!important` here, this would make further overrides harder.

xitij2000 replied:
Should we also fix this in the MFE so this hack is not needed? I guess that can be part of the theming epic, since the idea is to make things easier to theme and this seems a bit convoluted.

Agrendalath replied:
@xitij2000, good idea. I created [BB-7109](https://tasks.opencraft.com/browse/BB-7109).

Agrendalath replied:
@xitij2000, good idea. I created [BB-7109](https://tasks.opencraft.com/browse/BB-7109).

Agrendalath commented on lms/static/sass/theme-overrides.scss:
```diff
@@ -180,12 +182,6 @@
 }
 
 .content-wrapper {
-  // Customize the colors of the course navigation menu - bootstrap
-  h4 a:hover,
-  h3 a:hover {
-    color: $main-color !important; // Main color
-  }
-
```
The Course Outline page is now in the MFE.

Agrendalath commented on lms/static/sass/theme-overrides.scss:
```diff
@@ -275,74 +271,6 @@
-      @if variable-exists(account-settings-nav-color) {
-        // Customize the account settings nav menu color
-        color: $account-settings-nav-color !important;
-      }
-    }
-  }
-}
-
-.account-settings-sections .section {
-  .account-settings-section-body .u-field {
-    .field .u-field-link {
-      &,
-      &:hover {
-        color: $btn-primary-border-color;
-        border-color: $btn-primary-border-color;
-      }
-
-      &:hover {
-        color: #ffff;
-        border-color: $btn-primary-border-color;
-        background-color: $btn-primary-border-color;
-      }
-    }
-  }
-
-  .account-deletion-details .btn-outline-primary {
-    // Customize the account deletion button colors
-    & {
-      color: #b20610 !important;
-      border-color: #b20610 !important;
-    }
-
-    &:hover {
-      color: #fff !important;
-      border-color: #b20610 !important;
-      background-color: #b20610 !important;
-    }
-  }
-}
-
```
Account settings are now an MFE.

Agrendalath commented on paragon/_overrides.scss:
```diff
@@ -17,75 +17,34 @@ header {
     // it breaks the UI.
     height: calc(3.75rem + #{$header-border-top-width + $header-border-bottom-width});
   }
-  // Set the color of text in the header - course about page.
-  @if variable-exists(main-nav-color) {
-    span {
-      color: $main-nav-color;
-    }
-  }
```
The Course About page is in the legacy experience.

Agrendalath commented on paragon/_overrides.scss:
```diff
@@ -17,75 +17,34 @@ header {
-    }
-  }
-
-  .nav-links {
-    @if variable-exists(main-nav-link-color) {
-      // Set the color of the links in the dashboard
-      .main .nav-tab a,
-      .secondary .nav-item a,
-      .secondary .nav-item a span,
-      .toggle-user-dropdown span {
-        color: $main-nav-link-color !important;
-      }
-    }
-
-    @if variable-exists(main-nav-item-border-bottom-color) {
-      // Set the bottom border color of the links in the dashboard like 'Courses', 'Programs', 'Profile' etc.
-      .main .nav-tab a {
-        &.active {
-          border-bottom-color: $main-nav-item-border-bottom-color !important;
-          border-bottom-style: solid !important;
-        }
-      }
-    }
-
-    @if variable-exists(main-nav-item-hover-border-bottom-color) {
-      .main .nav-tab a {
-        &:not(.active):hover {
-          border-bottom-color: $main-nav-item-hover-border-bottom-color !important;
-          border-bottom-style: solid !important;
-        }
-      }
-    }
```
The dashboard is in the legacy experience.

Agrendalath commented on paragon/_overrides.scss:
```diff
@@ -17,75 +17,34 @@ header {
+    .user-dropdown,
+    // Customize the color of the links in the user dropdown menu in other MFEs
+    nav.secondary-menu-container {
+      .dropdown-menu a.dropdown-item {
```
This header seems to be inconsistent between MFEs. I didn't dig into this, though.

Agrendalath commented on paragon/_overrides.scss:
```diff
@@ -17,75 +17,34 @@ header {
     }
   }
 
-  .main .nav-tab a,
-  .secondary .nav-item a:not(.btn):not(.help-link),
-  .secondary .nav-item .toggle-user-dropdown {
-    // Set the color of the dropdown arrow of the user dropdowm menu
-    color: $link-color !important;
-  }
```
An arrow is not an `::after` element, not a button. Changing it does not match the username color.

Agrendalath commented on paragon/_overrides.scss:
```diff
@@ -17,75 +17,34 @@ header {
-      // Customize the register button on the home page header
-      a.register-btn {
+      // Customize the register button on the course page header
+      a.btn-outline-primary[href*="/register"] {
```
I don't like this approach, but it's the only reliable way I found to distinguish these buttons.

xitij2000 commented:
@Agrendalath I notice that we haven't versioned this repo with each release. I see us using the master branch in most places.  As soon as this merges, any current instances that are redeployed will pick up the changes. I see a few options to avoid this:
- We can update existing references to target the previous release branch, but that will involve a lot more PRs to all existing simple theme consumers, ocim, and config repos.
- We can merge this into a new branch for olive, and also merge-back the changes to the main branch after all instances are upgraded.
- We can hold off on merging this till fully prepared for the olive upgrade and merge it then. It will be a smaller window, but there could still be a few instances deployed in that duration that pick up the wrong changes. 

I'm favouring option 2 for now. Does that make sense? If so, could you push this to a new branch, and we can create another PR to merge as part of the Olive upgrade.

Agrendalath commented:
@xitij2000,
> We can merge this into a new branch for olive, and also merge-back the changes to the main branch after all instances are upgraded.

I like this option as a temporary solution. I'm okay with using either unique version names **or** unique version numbers, but we should stick to one in the long run. Double versioning (i.e., the release name **and** the version number) will get confusing pretty quickly. Nevertheless, semver would be a better fit for the npm package.

How about the following approach?
1. Create a `main` branch from the current `master` and mark it as a default one (AFAIK, all dependencies explicitly use `master`, not the default branch).
2. Update the README in the `master` branch to explain that `main` should be used from now on.
3. _Optionally_, add a "deprecated" field to `package.json` in the `master` branch. This doesn't seem supported in Node>v14 anymore, though.
4. Create a PR template explaining (as a comment) that the `master` branch should no longer be used.
5. Create a `0.1.0` release/tag from `main`.
6. Merge this into the `main` branch and create a `0.2.0` release/tag.
7. Delete the `master` branch at some point when we are sure nobody else is using it.

This way, we can:
1. Skip the `merge-back the changes to the main branch after all instances are upgraded.` step, which is very error-prone. If we do this, it can silently break the UI of a deployment where we missed its usage. This issue will not be easy to catch. On the other hand, deleting the `master` branch will break the builds, so we can catch this (and recreate the branch if we need an urgent deployment).
2. Make sure new changes are compatible with the current release. If we need to make adjustments directly to the Nutmeg version, we would apply fixes to the "upstream" (`main`) branch and then create a new branch from the `0.1.0` tag to prepare a `0.1.1` release. This would work similarly to what we're doing with any edx-platform dependencies.

What do you think?

xitij2000 commented:
> I like this option as a temporary solution. I'm okay with using either unique version names **or** unique version numbers, but we should stick to one in the long run. Double versioning (i.e., the release name **and** the version number) will get confusing pretty quickly. Nevertheless, semver would be a better fit for the npm package.

I agree. I think we should just continue the trend of using branches in the traditional `open-release/name.version` pattern, or perhaps `release/name` ref [this discussion](https://discuss.openedx.org/t/should-we-rename-the-release-branches/8827/2). 

I wasn't very concerned with the npm package version when the package was created, so "0.1.0" was just a placeholder, pretty much like it's always "0.1.0" for "edx-platform". This isn't designed to be released on npm, so the version number is just to satisfy the package.json format. However, if we want to start using the version field, I feel it'd be better to bump it to "1.0" and "2.0". 

I think your approach makes sense but, I just have one concern, which is marking the new `main` branch as default. Are you certain this won't make new deployments pull from that branch? If so, it's as good as merging to the current `master` branch. 

I'm not sure why merging back changes later would be error prone. The idea would be that from now on, we specify the branch for this repo just like we do for most others, so everywhere we upgrade to olive we'd use `release/olive` and when all instances are upgraded to olive, we can merge `release/olive` to `master`/`main`.

Agrendalath commented:
@xitij2000,
> or perhaps release/name ref [this discussion](https://discuss.openedx.org/t/should-we-rename-the-release-branches/8827/2).

Nice! I didn't see this one before.

> I think your approach makes sense but, I just have one concern, which is marking the new main branch as default. Are you certain this won't make new deployments pull from that branch? If so, it's as good as merging to the current master branch.

I should have mentioned this in my previous comment, as I was concerned about this too. There are three places I'm aware of that could affect this:
1. [Configuration playbook](https://github.com/openedx/configuration/blob/77eeab7/playbooks/roles/simple_theme/tasks/deploy.yml#L43-L43) - it uses `master` by default, though the `EDXAPP_COMPREHENSIVE_THEME_VERSION` variable evaluates to an empty string here, so [it might not even work](https://docs.ansible.com/ansible/latest/playbook_guide/playbooks_filters.html#providing-default-values) without this variable set.
2. [Ocim](https://gitlab.com/opencraft/dev/console/-/blob/master/opencraft/settings.py#L581), which sets the abovementioned variable to `master` if it's not defined.
3. [Grove](https://gitlab.com/opencraft/dev/grove-template/-/blob/main/instances/example/grove.yml#L3), which uses the `master` version as a default value. The version [needs to be defined](https://gitlab.com/opencraft/dev/grove/blob/3282b3258a8b8582d9a5193d04c7d5c18ac59570/tools-container/grove-cli/grove/instance.py#L344-348) and `npm install` uses an already cloned package.

> I'm not sure why merging back changes later would be error prone.

As long as we're 100% sure we have it explicitly defined everywhere, we should be fine. As described above, `master` is the default value in a few places.

> This isn't designed to be released on npm, so the version number is just to satisfy the package.json format.

It makes sense. So, I believe that the deciding factor between numbers vs branch names will be: "Do we want to automatically pull updates with each redeployment"? I believe we do, so branches will indeed be better here, so I changed the target branch of this PR and reverted the version bump.
I'm only worried that it will be the same as with the `opencraft-release/lilac.2` branch that nobody uses, so maybe we would be better changing the default branch to `release/olive`? For now, I just marked the `master` branch as read-only.


<br/>
Just thinking aloud: we could think about adding a GitHub action to prepare merge-back PRs in the future. Using the `release/olive` as the primary development resource, and doing `master` merge-backs manually is going to be tedious. We may eventually forget to move some changes, ending up with code drift.

xitij2000 commented:
> Just thinking aloud: we could think about adding a GitHub action to prepare merge-back PRs in the future. Using the `release/olive` as the primary development resource, and doing `master` merge-backs manually is going to be tedious. We may eventually forget to move some changes, ending up with code drift.

I agree, I think we can keep `main` targeting the latest release, and have new commits to `main` with changes cherry-picked to the relevant branches.

Agrendalath commented:
@xitij2000,
> I agree, I think we can keep main targeting the latest release, and have new commits to main with changes cherry-picked to the relevant branches.

Yep, that's also a good option. However, let's move it into a separate ticket because it would increase the scope here. We can create the `main` branch later.

Is there anything else we should do before we merge this?

xitij2000 reviewed (approved):
:+1: This is good to go! The theme seems to be applied on learning, profile and accounts. 

- [x] I tested this: tested on devstack
- [x] I read through the code
- [na] I checked for accessibility issues
- [na] Includes documentation
- [na] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.
